romancal/resample/resample.py

```diff
@@ -732,30 +734,40 @@ def gwcs_into_l3(model, wcsinfo):
     by indexing. This is fragile and will be a source of issues.
     """
     l3_wcsinfo = model.meta.wcsinfo
-    transform = wcsinfo.forward_transform
+    transform = wcs.forward_transform
 
     # Basic WCS info
     l3_wcsinfo.projection = "TAN"
+    l3_wcsinfo.pixel_shape = model.shape
     l3_wcsinfo.rotation_matrix = transform["pc_rotation_matrix"].matrix.value.tolist()
     l3_wcsinfo.dec_ref = transform.lat_6.value
     l3_wcsinfo.ra_ref = transform.lon_6.value
-    # l3_wcsinfo.x_ref = center of mosaic?
-    # l3_wcsinfo.y_ref = center of mosaic?
-    # l3_wcsinfo.s_region = from gwcs.bounding_box
-
-    # Mosaic info
-    # l3_wcsinfo.dec_center = from center of mosaic
-    # l3_wcsinfo.pixel_scale = ???
-    # l3_wcsinfo.pixel_scale_local = ???
-    # l3_wcsinfo.pixel_shape = image shape
-    # l3_wcsinfo.ra_center = from center of mosaic
-    # l3_wcsinfo.ra_corn1 = from bounding box
-    # l3_wcsinfo.ra_corn2 = from bounding box
-    # l3_wcsinfo.ra_corn3 = from bounding box
-    # l3_wcsinfo.ra_corn4 = from bounding box
-    # l3_wcsinfo.dec_corn1 = from bounding box
-    # l3_wcsinfo.dec_corn2 = from bounding box
-    # l3_wcsinfo.dec_corn3 = from bounding box
-    # l3_wcsinfo.dec_corn4 = from bounding box
-    # l3_wcsinfo.orientat = ???
-    # l3_wcsinfo.orientat_local = ???
+    l3_wcsinfo.x_ref = -transform.offset_0.value  # cdelt1
+    l3_wcsinfo.y_ref = -transform.offset_1.value  # cdelt2
+    l3_wcsinfo.pixel_scale = (transform.factor_3.value + transform.factor_4.value) / 2.0
+
+    world_center = wcs(*[v / 2.0 for v in model.shape])
+    world_center_plus = wcs(*[(v / 2.0) + 1 for v in model.shape])
+    world_delta = [w_plus - w for w_plus, w in zip(world_center_plus, world_center)]
+    l3_wcsinfo.ra_center = world_center[0]
+    l3_wcsinfo.dec_center = world_center[1]
+    l3_wcsinfo.pixel_scale_local = (world_delta[0] + world_delta[1]) / 2.0
+
+    footprint = utils.create_footprint(wcs, model.shape)
+    l3_wcsinfo.ra_corn1 = footprint[0][0]
+    l3_wcsinfo.ra_corn2 = footprint[1][0]
+    l3_wcsinfo.ra_corn3 = footprint[2][0]
+    l3_wcsinfo.ra_corn4 = footprint[3][0]
+    l3_wcsinfo.dec_corn1 = footprint[0][1]
+    l3_wcsinfo.dec_corn2 = footprint[1][1]
+    l3_wcsinfo.dec_corn3 = footprint[2][1]
+    l3_wcsinfo.dec_corn4 = footprint[3][1]
+    l3_wcsinfo.s_region = utils.create_s_region(footprint)
+
+    rm = l3_wcsinfo.rotation_matrix
+    xrot = abs(math.atan2(rm[1][0], rm[0][0]))
```

Comment:
Let's compute orientat_local as c_center.position_angle(c_up) using the SkyCoord created above.  We'll define it as the position angle relative to north of the Y axis of the image.

For orientat what you have could be okay, but I'm leery about assuming transform.pc_rotation_matrix exists?

@nden , conceptually we want to fill out these fields as appropriate for a tangent plane projection.  But the gWCS object can support ~arbitrary transformations and one could imagine stuffing in an extra no-op layer into the gWCS object so that reaching into the guts of the gWCS and expecting the parameters to be the right ones for the tangent plane projection seems dangerous.  It's probably true that we could get all of the "local" values only using the forward transformation and local differences.  We could also get all of the projection center values if we're allowed to evaluate the forward transformation at x_ref, y_ref, which will be outside the image but everything will be well defined there; i.e., the only reason we couldn't evaluate it there would be if we set the bounding box to explicitly prevent that.  The only thing I don't know how to get would be one of the ra/dec reference or the x/y reference point, which conceptually only exists in a tangent plane projection and so I don't know how to ask the generic gWCS object for it without supposing a form.  Any ideas there?

Reply:
I think I addressed this above.